File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrVirtualFilesUIExtension.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Platform.Client.UI;
using Tessa.Extensions.Platform.Server.Cards;
using Tessa.Platform.Conditions;
using Tessa.Platform.Placeholders;
using Tessa.Platform.Runtime;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;
using Unity;

namespace Tessa.Extensions.Default.Client.UI
{
    public sealed class KrVirtualFilesUIExtension : CardUIExtension
    {
        #region Fields

        private readonly ISession session;
        private readonly IUnityContainer unityContainer;
        private readonly ICardRepairManager cardRepairManager;
        private readonly ICardMetadata cardMetadata;
        private readonly IConditionTypesProvider typesProvider;
        private readonly IPlaceholderManager placeholderManager;

        #endregion

        #region Constructors

        public KrVirtualFilesUIExtension(
            ISession session,
            IUnityContainer unityContainer,
            ICardRepairManager cardRepairManager,
            ICardMetadata cardMetadata,
            IConditionTypesProvider typesProvider,
            IPlaceholderManager placeholderManager)
        {
            this.session = session;
            this.unityContainer = unityContainer;
            this.cardRepairManager = cardRepairManager;
            this.cardMetadata = cardMetadata;
            this.typesProvider = typesProvider;
            this.placeholderManager = placeholderManager;
        }

        #endregion

        #region Base Overrides

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            var cardModel = context.Model;
            var card = cardModel.Card;

            var conditionContext = new ConditionsUIContext(
                session,
                unityContainer,
                cardRepairManager,
                cardMetadata,
                typesProvider,
                placeholderManager);

            conditionContext.Initialize(context.Model);

            card.Sections.GetOrAddTable("KrVirtualFileVersions").Rows.ItemChanged += (s, e) =>
            {
                if (e.Action == ListStorageAction.Insert)
                {
                    EventHandler<CardRowStateEventArgs> stateChanged = null;
                    stateChanged = (item, args) =>
                    {
                        if (item is CardRow row)
                        {
                            row["FileVersionID"] = row.RowID;
                            row.StateChanged -= stateChanged;
                        }
                    };
                    e.Item.StateChanged += stateChanged;
                }
            };

            if (cardModel.Controls.TryGet("CompileButton", out var compileButtonObj)
                && compileButtonObj is ButtonViewModel button)
            {
                button.CommandClosure.Execute = CompileAsync;
            }

            async void CompileAsync(object obj)
            {
                var storeInfo = new Dictionary<string, object>(StringComparer.Ordinal);
                storeInfo.SetCompileMark();

                await UIContext.Current.CardEditor.SaveCardAsync(
                    UIContext.Current,
                    storeInfo);
            }
        }

        #endregion
    }
}

File: WebClient SDK/src/default/ui/krVirtualFilesUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { ConditionsUIContext } from 'tessa/ui/conditions';
import { CardTableType, CardSectionType, CardRowStateChangedEventArgs, systemKeyPrefix } from 'tessa/cards';
import { DotNetType, createTypedField } from 'tessa/platform';
import { ButtonViewModel } from 'tessa/ui/cards/controls';
import { IStorage } from 'tessa/platform/storage';
import { UIContext } from 'tessa/ui';

export class KrVirtualFilesUIExtension extends CardUIExtension {

  public shouldExecute(context: ICardUIExtensionContext) {
    return context.card.typeId === '81250a95-5c1e-488c-a423-106e7f982c6b'; // KrVirtualFileTypeID
  }

  public initialized(context: ICardUIExtensionContext) {
    const cardModel = context.model;
    const card = cardModel.card;

    const conditionContext = new ConditionsUIContext();
    conditionContext.initialize(context.model);

    const virtualSection = card.sections.getOrAdd('KrVirtualFileVersions');
    virtualSection.type = CardSectionType.Table;
    virtualSection.tableType = CardTableType.Collection;

    virtualSection.rows.collectionChanged.add(e => {
      if (e.added.length > 0) {
        for (let item of e.added) {
          item.stateChanged.add(this.rowStateChanged);
        }
      }
    });

    const button = cardModel.controls.get('CompileButton') as ButtonViewModel;
    if (button) {
      button.onClick = this.compile;
    }
  }

  private rowStateChanged = (e: CardRowStateChangedEventArgs) => {
    e.row.set('FileVersionID', e.row.rowId, DotNetType.Guid);
    e.row.stateChanged.remove(this.rowStateChanged);
  }

  private compile = async () => {
    const storeInfo: IStorage = {};
    storeInfo[systemKeyPrefix + 'Compile'] = createTypedField(true, DotNetType.Boolean);

    await UIContext.current.cardEditor!.saveCard(UIContext.current, storeInfo);
  }

}